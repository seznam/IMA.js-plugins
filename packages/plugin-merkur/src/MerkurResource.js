import { Cache, HttpAgent } from '@ima/core';

export default class MerkurResource {
  static get $dependencies() {
    return [HttpAgent, Cache];
  }

  /**
   *
   * @param {HttpAgent} http
   * @param {Cache} cache
   */
  constructor(http, cache) {
    /**
     * @type {HttpAgent}
     */
    this._http = http;

    /**
     * @type {Cache}
     */
    this._cache = cache;
  }

  /**
   * Make API call to defined url and returns response from server.
   *
   * @param {string} url
   * @param {Object<string, *>} data
   * @param {Object<string, *>} options
   * @return {Promise<Response>} response
   */
  async get(url, data, options = {}) {
    if (!data.containerSelector) {
      throw new Error(
        'The containerSelector property must be set in data argument.'
      );
    }

    let cloneData = Object.assign({}, data);
    const { containerSelector, slot = {} } = cloneData;

    delete cloneData.containerSelector;
    delete cloneData.slot;

    options = this._addDefaultRequestOptions(options);
    const response = await this._http[options.method](url, cloneData, options);

    this._removeHTMLFromCache(url, cloneData, options);

    if (response && response.body) {
      response.body.containerSelector = containerSelector;

      // Deprecated, left for backwards compatibility
      if (response.body.props) {
        response.body.props.containerSelector = containerSelector;
      }

      if (Object.keys(slot).length > 0 && response.body.slot) {
        Object.keys(slot).forEach(slotName => {
          if (response.body.slot[slotName]) {
            response.body.slot[slotName].containerSelector =
              slot[slotName].containerSelector;
          }
        });
      }
    }

    return response;
  }

  _addDefaultRequestOptions(options) {
    const cloneOptions = Object.assign({}, options);

    if (cloneOptions.cache === undefined) {
      cloneOptions.cache = true;
    }

    cloneOptions.method = cloneOptions.method || 'get';
    cloneOptions.ttl = options.ttl || 60000;

    return cloneOptions;
  }

  _removeHTMLFromCache(url, data, options) {
    const cacheKey = this._http.getCacheKey(options.method, url, data);

    if (this._cache.has(cacheKey)) {
      let cacheValue = this._cache.get(cacheKey);

      if ('html' in cacheValue.body) {
        delete cacheValue.body.html;

        if ('slot' in cacheValue.body) {
          Object.values(cacheValue.body.slot).forEach(slot => {
            if ('html' in slot) {
              delete slot.html;
            }
          });
        }

        this._cache.set(cacheKey, cacheValue, options.ttl);
      }
    }
  }
}
