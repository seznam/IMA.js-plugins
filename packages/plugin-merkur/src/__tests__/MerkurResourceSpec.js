import MerkurResource from '../MerkurResource';
import { Cache, HttpAgent } from '@ima/core';
import { toMockedInstance, setGlobalMockMethod } from 'to-mock';

setGlobalMockMethod(jest.fn);

describe('MerkurResource clase', () => {
  let merkurResource = null;
  let http = null;
  let cache = null;

  let url = 'http://example.com/widget';
  let data = {
    containerSelector: '.some-class'
  };
  let options = {};

  let body = {
    name: 'my-widget',
    version: '0.0.1',
    props: {},
    state: {
      counter: 0
    },
    assets: [
      {
        name: 'polyfill.js',
        type: 'script',
        source: {
          es5: 'http://localhost:4444/static/es5/polyfill.31c5090d8c961e43fade.js'
        },
        test: 'return window.fetch'
      },
      {
        name: 'widget.js',
        type: 'script',
        source: {
          es9: 'http://localhost:4444/static/es9/widget.6961af42bfa3596bb147.js',
          es5: 'http://localhost:4444/static/es5/widget.31c5090d8c961e43fade.js'
        },
        attr: {
          async: true,
          'custom-attribute': 'foo'
        }
      },
      {
        name: 'widget.css',
        type: 'stylesheet',
        source:
          'http://localhost:4444/static/es9/widget.814e0cb568c7ddc0725d.css'
      }
    ],
    html: '<div></div>'
  };

  beforeEach(() => {
    http = toMockedInstance(HttpAgent, {
      get() {
        return { body, headers: {}, params: data, status: 200, cached: true };
      }
    });
    cache = toMockedInstance(Cache);
    merkurResource = new MerkurResource(http, cache);
  });

  it('should throw error for not defined containerSelector', async () => {
    await expect(merkurResource.get()).rejects.toThrow(
      "Cannot read property 'containerSelector' of undefined"
    );
  });

  it('should return response from widget API with containerSelector', async () => {
    let response = await merkurResource.get(url, data, options);

    expect(response.body).toMatchSnapshot();
  });

  it('should add default options to widget API', async () => {
    http.get = jest.fn();

    await merkurResource.get(url, data, options);

    expect(http.get.mock.calls[0][2]).toMatchSnapshot();
  });

  it('should remove containerSelector from data', async () => {
    http.get = jest.fn();

    await merkurResource.get(url, data, options);

    expect(http.get.mock.calls[0][1]).toMatchSnapshot();
  });
});
