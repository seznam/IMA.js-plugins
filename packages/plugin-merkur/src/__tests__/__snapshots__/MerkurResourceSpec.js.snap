// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MerkurResource class default usage should return response from widget API with containerSelector set to both props and widget 1`] = `
Object {
  "assets": Array [
    Object {
      "name": "polyfill.js",
      "source": Object {
        "es5": "http://localhost:4444/static/es5/polyfill.31c5090d8c961e43fade.js",
      },
      "test": "return window.fetch",
      "type": "script",
    },
    Object {
      "attr": Object {
        "async": true,
        "custom-attribute": "foo",
      },
      "name": "widget.js",
      "source": Object {
        "es5": "http://localhost:4444/static/es5/widget.31c5090d8c961e43fade.js",
        "es9": "http://localhost:4444/static/es9/widget.6961af42bfa3596bb147.js",
      },
      "type": "script",
    },
    Object {
      "name": "widget.css",
      "source": "http://localhost:4444/static/es9/widget.814e0cb568c7ddc0725d.css",
      "type": "stylesheet",
    },
  ],
  "containerSelector": ".some-class",
  "html": "<div></div>",
  "name": "my-widget",
  "props": Object {
    "containerSelector": ".some-class",
  },
  "state": Object {
    "counter": 0,
  },
  "version": "0.0.1",
}
`;

exports[`MerkurResource class slot usage should append container selectors to defined slot 1`] = `
Object {
  "body": Object {
    "assets": Array [
      Object {
        "name": "polyfill.js",
        "source": Object {
          "es5": "http://localhost:4444/static/es5/polyfill.31c5090d8c961e43fade.js",
        },
        "test": "return window.fetch",
        "type": "script",
      },
      Object {
        "attr": Object {
          "async": true,
          "custom-attribute": "foo",
        },
        "name": "widget.js",
        "source": Object {
          "es5": "http://localhost:4444/static/es5/widget.31c5090d8c961e43fade.js",
          "es9": "http://localhost:4444/static/es9/widget.6961af42bfa3596bb147.js",
        },
        "type": "script",
      },
      Object {
        "name": "widget.css",
        "source": "http://localhost:4444/static/es9/widget.814e0cb568c7ddc0725d.css",
        "type": "stylesheet",
      },
    ],
    "containerSelector": ".some-class",
    "html": "<div></div>",
    "name": "my-widget",
    "props": Object {
      "containerSelector": ".some-class",
    },
    "slot": Object {
      "headline": Object {
        "containerSelector": ".headline",
        "html": "<div></div>",
        "name": "headline",
      },
    },
    "state": Object {
      "counter": 0,
    },
    "version": "0.0.1",
  },
  "cached": true,
  "headers": Object {},
  "params": Object {
    "containerSelector": ".some-class",
  },
  "status": 200,
}
`;
